### PR TITLE
Normative: Allow enumerating the own keys of a module namespace

### LINKED ISSUE #1209: Object.keys throws for uninitialized module namespace object
Found in Node [here](https://github.com/nodejs/node/pull/20962) and filed as a V8 bug [here](https://bugs.chromium.org/p/v8/issues/detail?id=7780). It looks like `Object.keys` will throw an error when provided an uninitialized module namespace object while `Object.getOwnPropertyNames` won't.

Comments from the V8 issue:

> so i was talking with ljharb about this and we came to a different conclusion about this behaviour.
>
>Since Object.keys (and Reflect.ownKeys, for..in, all the things broken by this upgrade) don't actually expose the value, they shouldn't throw. basically everything that doesn't expose the value, or a getter/setter, should work. anything that's static and won't change after evaluation should be always accessible.

And

> See https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects.
>
>[[GetOwnProperty]] throws because [[Get]] throws:
>
>  Let value be ? O.[[Get]](P, O).
>
>[[Get]] throws because GetBindingValue throws:
>
>  Return ? targetEnvRec.GetBindingValue(binding.[[BindingName]], true).
>
> GetBindingValue throws because the binding, which exists, is still uninitialized:
>https://tc39.github.io/ecma262/#sec-module-environment-records-getbindingvalue-n-s

The bug is that `Object.keys` should not error just as `Object.getOwnPropertyNames` does not error.

\cc @ljharb

### PR DESCRIPTION
**Pull request by devsnek**
Fixes #1209

Tests: https://github.com/tc39/test262/pull/1693

### CODE REVIEW

**Comment by devsnek** [comment]
It should be mentioned that while there is no even testing surface for getting an unevaluated module namespace, Node.js exposes such an interface, which is how this behaviour was identified in the first place: https://nodejs.org/api/vm.html#vm_module_namespace

File: spec.html [hunk 1]
```diff
@@ -8574,10 +8574,10 @@ <h1>[[Delete]] ( _P_ )</h1>
...
-          1. Let _exports_ be a copy of _O_.[[Exports]].
-          1. Let _symbolKeys_ be ! OrdinaryOwnPropertyKeys(_O_).
-          1. Append all the entries of _symbolKeys_ to the end of _exports_.
-          1. Return _exports_.
+          1. Let _keys_ be a copy of _O_.[[Exports]].
+          1. For each own property key _P_ of _O_ that is a Symbol, in ascending chronological order of property creation, do
+            1. Add _P_ as the last element of _keys_.
+          1. Return _keys_.
         </emu-alg>
       </emu-clause>
```
Comment (jdalton) [inline spec.html]: `OrdinaryOwnPropertyKeys` looks to do more than symbol properties. Is that accounted for?

Comment (devsnek) [reply spec.html]: module namespace exotic objects cannot have non-symbol properties that aren't from [[Exports]]

**Comment by bakkot** [comment]
I don't think this patch is sufficient. Because Object.keys needs to return only _enumerable_ properties, it calls `[[GetOwnProperty]]` for each property (in step 4.a.i of [`EnumerableOwnPropertyNames`](https://tc39.github.io/ecma262/#sec-enumerableownpropertynames), which it [calls](https://tc39.github.io/ecma262/#sec-object.keys)).

But `[[GetOwnProperty]]` on module namespace exotic objects throws when the module is not initialized. So `Object.keys` will still throw after this patch unless the module being reflected on exports no bindings - not really an improvement, I think.

**Comment by ljharb** [comment]
So then it seems like we need instead to ensure the TDZ is on the value, not the descriptor?

**Comment by bakkot** [comment]
Or something like that, yeah. I don't think the spec mechanics currently allow precisely that: records [are just](https://tc39.github.io/ecma262/#sec-list-and-record-specification-type) wrappers around one or more values, and can't really express "it's an error to try to access this field of the record". (That why `[[...]]` access just gives a value, not a completion record which needs to be unwrapped with `!` or `?`.)

**Comment by devsnek** [comment]
This needs a different approach.

**Comment by bterlson** [comment]
IMO we shouldn't be in the business of adding MOP complexity to deal with objects where we know the properties but not the values. I want to preserve the notion that "if a property exists, it has a descriptor (or can have one synthesized)."

The bug here is that Object.getOwnPropertyNames doesn't throw. Making these reflection things "kinda work" doesn't seem super well motivated to me. I'd also be ok if we want to represent this as getters which throw when the value isn't yet available. Let's not create new kinds of exotics if we can help it.

**Pull request closed without merging**